experimental/oak_async/src/executor.rs

```diff
@@ -0,0 +1,181 @@
+            reader_id,
+            handle.handle
+        );
+        self.waiting_readers.insert(
+            reader_id,
+            WaitingReader {
+                handle: handle.handle,
+                waker: waker.clone(),
+            },
+        );
+    }
+
+    pub fn remove_waiting_reader(&mut self, reader_id: ReaderId) {
+        trace!("Remove waiting reader {}", reader_id);
+        if self.waiting_readers.remove(&reader_id).is_none() {
+            // This is usually not an error. If a Future is dropped as a result of it being woken up
+            // and then resolving, we expect the reader_id to not be present in the waiting set.
+            debug!(
+                "Attempted to remove waiting reader {}, but no such reader was in the waiting set",
+                reader_id
+            )
+        }
+    }
+
+    pub fn new_id(&mut self) -> ReaderId {
+        let id = self.next_reader_id;
+        // Collisions are technically possible, but only if one read is very slow and many async
+        // reads (2^64) are requested in quick succession.
+        self.next_reader_id = self.next_reader_id.wrapping_add(1);
+        id
+    }
+
+    pub fn none_waiting(&self) -> bool {
+        self.waiting_readers.is_empty()
+    }
+
+    /// Returns all handles for which reads are pending, and the id of one of the readers for that
+    /// channel.
+    ///
+    /// The `Vec`s returned are guaranteed to have the same length, and a handle at index `i`
+    /// corresponds to the reader id at index `i`.
+    pub fn pending_handles(&self) -> (Vec<ReadHandle>, Vec<ReaderId>) {
+        self.waiting_readers
+            .iter()
+            .map(|(reader_id, waiting_reader)| (waiting_reader.handle, *reader_id))
+            .collect::<HashMap<Handle, ReaderId>>()
+            .into_iter()
+            .map(|(handle, reader_id)| (ReadHandle { handle }, reader_id))
+            .unzip()
+    }
+
+    /// Remove a reader from the waiting set and wake it
+    pub fn wake_reader(&mut self, reader_id: ReaderId) {
+        self.waiting_readers
+            .remove(&reader_id)
+            .unwrap()
+            .waker
+            .wake();
+    }
+}
+
+/// Runs the given closure, providing it with a handle to the current executor
+pub fn with_executor<F: FnOnce(&mut Executor) -> R, R>(f: F) -> R {
+    EXECUTOR.with(|executor| f(&mut executor.borrow_mut()))
+}
+
+/// Block the current thread until the provided `Future` has been `poll`ed to completion.
+///
+/// Returns `Err(_)` if the call to `wait_on_channels` fails.
+pub fn block_on<F: Future + 'static>(f: F) -> Result<F::Output, OakStatus> {
+    let mut pool = LocalPool::new();
+    let spawner = pool.spawner();
+    let main_join_handle = spawner
+        .spawn_local_with_handle(f)
+        .expect("Failed to spawn main future");
+    loop {
+        // Poll futures in the pool until none of them can make any progress.
+        pool.run_until_stalled();
```

Comment:
To confirm: this will prioritize futures that can make progress without blocking, before even starting to block on any handle-based one, right? If so, I guess this means that if there are many futures, each expecting to read some data and then do a lot of computation on the result, the first one will be polled to wait on the channel, perhaps receive its result immediately (before others get a chance of being scheduled) and then the CPU-bound computation will continue for a long time, until it stalls again, at which point the second IO-bound future will be polled for the first time. Is this correct?

Reply:
Yes, this is correct. `wait_on_channels` is only called if at one point no single future can make any progress, i.e. all of them have returned `Poll::Pending` and there is nothing else we can do.